src/python/pants/engine/internals/graph_test.py

```diff
@@ -153,6 +156,52 @@ class MockTargetGenerator(TargetFilesGenerator):
     moved_fields = (MockDependencies, Tags, ResolveField)
 
 
+class MockAlsoGeneratedTarget(Target):
+    alias = "also_generated"
+    core_fields = (MockDependencies, Tags, MockSingleSourceField, ResolveField)
```

Comment:
I'd like to see what happens if the generated targets have a disjoint set of fields. Say, if this target doesn't have the `resolve` field, but instead has another field that `generated` does not have. Can we provide both of these fields to the generator, and it will sort through them to apply to the correct target?

Reply:
Right now, everything from the target generators core fields is propagated to the template and the same template is used for all types. My reasoning for doing it this way is that since rule authors are ultimately responsible for propagating the fields downstream to the generated targets, we should just provide them everything and let them decide how the targets are generated.

Actually what I just realized is currently we make multiple generate targets request but do not specify in the request which target type each template is for. Instead we should creating a single request for all types that breaks down the templates by type. This is kinda tricky because I think the ideal solution would be we provide a template based only off what's provided by the generator and then another dict of templates broken down by type which also provide defaults. However, this would break backwards compatibility. The obvious solution would be to include defaults in the "untyped " form of the template for single generated class type target generators and exclude them for multi-type generators but it's not exactly the cleanest design.

Reply:
Yea, I think it's good enough to let the target generator take care of combing the fields to the respective targets it generates. That is, pass in all defaults for all targets in a single request. (if we split this up per generated target type for the sake of defaults, that would be the only way to pass different values per target type, which doesn't really make sense as we can't do that directly on the target generator itself.)

Reply:
What if I just add a separate field for defaults by type, don't apply defaults to the template for multi-type target generators, and then rely on rule authors to handle applying defaults correctly?

Reply:
For consistency, I'd prefer that the target generator doesn't have the ability to pick apart defaults.
Consider this example:
```python
__defaults__({tgt_a: dict(a=1), tgt_b: dict(b=2)})

# if we generate `tgt_a` and `tgt_b`, we should also support their fields:
my_generator(a=3, b=4)
```

The generator must support either copying or moving the fields of the generated target types it claims to generate (as being part of it's interface). If we pass in defaults on the side to the generator, we open up a lot of complexity for the generator implementation to deal with applying defaults, which I'd rather we take care of in pants core, as well as open up for the possibility of the generator to not accept all the fields directly, relying on defaults for the only way to provide certain values. That to me is an undesirable split.
(we may still generate targets not part of the interface, but then we also won't get any defaults for these target types.. perhaps that is a separate issue to look at how to get at the defaults from a plugin?)

Reply:
What about the following use case though?

```
__defaults__({tgt_a: dict(skip_lint=True), tgt_b: dict(skip_lint=False)})

# if we generate `tgt_a` and `tgt_b`, we should also support their fields:
my_generator(a=3, b=4)
```

Is the solution here to have 2 separate skip lint fields? Trying to recall how separate fields with the same alias would be handled.

Reply:
That's exactly the point I try to avoid supporting. If you have a target generator like that, the _only_ way to provide values for the targets are via defaults, which I find is a mis-design. The defaults should only be used to provide values to many targets to reduce repetition and boilerplate. 

If we want to support target generators that generate multiple types of targets with a dis-joint set of fields, we should support providing field values for them on the target generator. Maybe `overrides` could be leveraged for this purpose, if we really need it?

Reply:
Or... I'm too pure here. What if we have target generators that doesn't take any inputs at all (or just a fraction of the support fields of the targets it generates), and we should rely on defaults to "override" or provide values for all those generated targets.

Reply:
Okay I think I maybe see a path forward here. Rule authors can provide separate fields on the target generators e.g. `skip_lint_a` and `skip_lint_b` and then if we want to set defaults for those we can use `__defaults__` on the target generator itself, i.e. `__defaults__({generator: dict(skip_lint_a=True, skip_lint_b=False)})`. It's certainly not the simplest implementation but this isn't exactly the simplest use case and I think it lets us avoid doing something we might regret later 😅.

Reply:
What I can also do is add a class attribute to the target generators defining behavior for conflicting `__defaults__`. We could even roll that + the generated target classes tuple into one class. I think that makes a lot of sense actually as it lets rule authors define how they want their templates generated in the case of conflicts.